libs/test-utils/src/mock_function.test.ts

```diff
@@ -0,0 +1,218 @@
+      .expectCallWith('a', 1, true, null, undefined, { foo: 'bar' }, [1, 2])
+      .returns('success');
+    expect(
+      funcMock('a', 1, true, null, undefined, { foo: 'bar' }, [1, 2])
+    ).toBe('success');
+    expect(() =>
+      funcMock('a', 1, true, null, undefined, { foo: 'bar' }, [1, 2])
+    ).toThrowErrorMatchingInlineSnapshot(
+      `"Unexpected call to mock function: funcWithManyTypes('a', 1, true, null, undefined, { foo: 'bar' }, [ 1, 2 ])"`
+    );
+    funcMock.reset();
+    funcMock
+      .expectCallWith('a', 1, true, null, undefined, { foo: 'bar' }, [1, 2])
+      .returns('success');
+    expect(() =>
+      funcMock('a', 1, true, null, undefined, { foo: 'wrong' }, [1, 2])
+    ).toThrowErrorMatchingInlineSnapshot(`
+      "Mismatched call to mock function:
+      Expected: funcWithManyTypes('a', 1, true, null, undefined, { foo: 'bar' }, [ 1, 2 ])
+      Actual: funcWithManyTypes('a', 1, true, null, undefined, { foo: 'wrong' }, [ 1, 2 ])"
+    `);
+  });
+
+  it('enforces correct types', () => {
+    const addMock = mockFunction<typeof add>('add');
+    // @ts-expect-error - wrong argument type
+    addMock.expectCallWith('1', 2).returns(3);
+    // @ts-expect-error - wrong return type
+    addMock.expectCallWith(1, 2).returns('3');
+  });
+
+  it('supports async functions', async () => {
+    // eslint-disable-next-line @typescript-eslint/require-await
+    async function addAsync(num1: number, num2: number): Promise<number> {
+      throw new Error('Not implemented');
+    }
+    const addMock = mockFunction<typeof addAsync>('addAsync');
+    // @ts-expect-error - can't use returns with async functions
+    addMock.expectCallWith(1, 2).returns(3);
+    addMock.expectCallWith(1, 2).resolves(3);
+    expect(await addMock(1, 2)).toBe(3);
+  });
+
+  it('supports mocking an exception', () => {
+    const addMock = mockFunction<typeof add>('add');
+    const error = new Error('Mock error');
+    addMock.expectCallWith(1, 2).throws(error);
+    expect(() => addMock(1, 2)).toThrow(error);
+    expect(console.error).not.toHaveBeenCalled();
+  });
+
+  it('assertComplete errors if not all expected calls are used', () => {
+    const addMock = mockFunction<typeof add>('add');
+    addMock.expectCallWith(1, 2).returns(3);
+    addMock.expectCallWith(2, 2).returns(4);
+    addMock(1, 2);
+    expect(() => addMock.assertComplete()).toThrowErrorMatchingInlineSnapshot(`
+      "Mismatch between expected mock function calls and actual mock function calls:
+
+      Call #0
+      Expected: add(1, 2)
+      Actual: add(1, 2)
+
+      Call #1
+      Expected: add(2, 2)
+      Actual: <none>"
+    `);
+  });
+
+  it('assertComplete errors if there are unexpected calls', () => {
+    const addMock = mockFunction<typeof add>('add');
+    addMock.expectCallWith(1, 2).returns(3);
+    expect(addMock(1, 2)).toBe(3);
+    expect(() => addMock(1, 3)).toThrow();
+    expect(() => addMock.assertComplete()).toThrowErrorMatchingInlineSnapshot(`
```

Comment:
So even though `addMock(1, 3)` will already throw and print an error, we'll get a more holistic summary from `addMock.assertComplete()` - nice!

Reply:
Yes, I've found it quite useful so far!